src/Microsoft.AspNetCore.WebHooks.Receivers/ApplicationModels/WebHookMetadataProvider.cs

```diff
@@ -0,0 +1,270 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+using Microsoft.AspNetCore.WebHooks.Metadata;
+using Microsoft.AspNetCore.WebHooks.Properties;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.WebHooks.ApplicationModels
+{
+    /// <summary>
+    /// <para>
+    /// An <see cref="IApplicationModelProvider"/> implementation that adds <see cref="IWebHookMetadata"/>
+    /// references to WebHook <see cref="ActionModel"/>s. Metadata is stored in <see cref="ActionModel.Properties"/>
+    /// and used in <see cref="WebHookModelBindingProvider"/> and <see cref="WebHookRoutingProvider"/>.
+    /// </para>
+    /// <para>
+    /// Detects missing and duplicate <see cref="IWebHookMetadata"/> services.
+    /// </para>
+    /// </summary>
+    public class WebHookMetadataProvider : IApplicationModelProvider
+    {
+        private readonly IReadOnlyList<IWebHookBindingMetadata> _bindingMetadata;
+        private readonly IReadOnlyList<IWebHookEventMetadata> _eventMetadata;
+        private readonly IReadOnlyList<IWebHookRequestMetadataService> _requestMetadata;
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Instantiates a new <see cref="WebHookMetadataProvider"/> with the given <paramref name="metadata"/>.
+        /// </summary>
+        /// <param name="metadata">The collection of <see cref="IWebHookMetadata"/> services.</param>
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+        public WebHookMetadataProvider(
+            IEnumerable<IWebHookMetadata> metadata,
+            ILoggerFactory loggerFactory)
+        {
+            _bindingMetadata = metadata.OfType<IWebHookBindingMetadata>().ToArray();
+            _eventMetadata = metadata.OfType<IWebHookEventMetadata>().ToArray();
+            _requestMetadata = metadata.OfType<IWebHookRequestMetadataService>().ToArray();
```

Comment:
This seems like it's doing an end-run around DI? Why aren't these things just registered in DI with the approriate types?

Reply:
Done this way for flexibility. All of the metadata interfaces inherit from `IWebHookMetadata` and that's how the implementations should be registered. This mechanism allows specific receivers to register a single metadata object once or multiple per-interface instances however much they like.

If I instead had the receivers register using the specific interfaces, could still have one object per receiver that does the implementation. But, the registration would get messy and multiple classes would be aware of the chosen implementations -- the implementation and the extension methods doing registration.